### PR TITLE
Register and launch single task execution

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
This change implements register_and_launch for single task executions featuring containerless tasks.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
The use case for this feature is in-notebook task execution. Because only containerless tasks do not need to update an image during task iteration, only those are supported for register_and_launch.

## Tracking Issue
https://github.com/lyft/flyte/issues/256

## Follow-up issue
https://github.com/lyft/flyte/issues/221

### CODE REVIEW

File: flytekit/common/tasks/presto_task.py [hunk 1]
```diff
@@ -122,6 +127,21 @@ def __call__(self, *args, **kwargs):
             *args, **kwargs
         )
 
+    # Override method in order to set the implicit inputs
+    def _python_std_input_map_to_literal_map(self, inputs):
+        """
```
Comment (wild-endeavor) [inline flytekit/common/tasks/presto_task.py:132]: Can you explain what the stuff in the __call__ function does and why it's not enough please, for me and for posterity to understand?

Comment (katrogan) [reply flytekit/common/tasks/presto_task.py:132]: ordinarily we use `__call__` to create an SDK node, which is formerly the only the way a task was transformed for execution.

but when we call `launch` on a task to trigger a single task execution [here](https://github.com/lyft/flytekit/blob/master/flytekit/common/mixins/launchable.py#L32) we don't use this code path and the implicit inputs are never added

File: flytekit/models/common.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def __str__(self):
         return self.verbose_string()
 
     def __hash__(self):
-        return hash(self.to_flyte_idl().SerializeToString())
+        return hash(self.to_flyte_idl().SerializeToString(deterministic=True))
```
Comment (wild-endeavor) [inline flytekit/models/common.py:59]: @matthewphsmith is there a reason this wasn't here initially?

File: flytekit/common/tasks/task.py [hunk 3]
```diff
@@ -268,6 +273,58 @@ def _python_std_input_map_to_literal_map(self, inputs):
             for k, v in _six.iteritems(self.interface.inputs)
         })
 
+    def _produce_deterministic_version(self, version=None):
+        """
+        :param Text version:
+        :return Text:
+        """
+
+        if self.container is not None:
```
Comment (wild-endeavor) [inline flytekit/common/tasks/task.py:282]: This won't work for Ketan's [raw container tasks](https://github.com/lyft/flytekit/pull/107/files#diff-0b4aabc3c6488fa90c0e7f3afca65ae7R214) btw.  How about a whitelist of allowed task types (strings) instead?

File: flytekit/common/tasks/task.py [hunk 4]
```diff
@@ -268,6 +273,58 @@ def _python_std_input_map_to_literal_map(self, inputs):
...
+    def register_and_launch(self, project, domain, name=None, version=None, inputs=None):
+        """
+        :param Text project: The project in which to register and launch this task.
+        :param Text domain: The domain in which to register and launch this task.
+        :param Text name: The name to give this task.
+        :param Text version: The version in which to register this task
+
+        :param dict[Text, Any] inputs: A dictionary of Python standard inputs that will be type-checked, then compiled
+            to a LiteralMap.
+        """
```
Comment (wild-endeavor) [inline flytekit/common/tasks/task.py:306]: Add the return type?

File: flytekit/common/tasks/task.py [hunk 5]
```diff
@@ -268,6 +273,58 @@ def _python_std_input_map_to_literal_map(self, inputs):
...
+        self.validate()
+        version = self._produce_deterministic_version(version)
+
+        if name is None:
+            try:
+                self.auto_assign_name()
+                generated_name = self._platform_valid_name
```
Comment (wild-endeavor) [inline flytekit/common/tasks/task.py:313]: can remove the _ if you want

File: flytekit/common/tasks/task.py (hunk 3 above)
Comment (kumare3) [reply flytekit/common/tasks/task.py:282]: @wild-endeavor raw container tasks are not task type. You just have to check a different, thing, as this proposal
```suggestion
        if self.container is not None and self.container. data_config is None:
```

File: flytekit/common/tasks/task.py (hunk 4 above)
Comment (kumare3) [reply flytekit/common/tasks/task.py:306]: I want to start using python3 typing.

File: flytekit/models/common.py (hunk 2 above)
Comment (kumare3) [reply flytekit/models/common.py:59]: ohh wow this looks like a bug? But, we never really use this right, we always use the git-sha as the version

**Comment by kumare3** [comment]
Is there an example execution of single task from command line / jupyter?

File: flytekit/common/tasks/presto_task.py (hunk 1 above)
Comment (EngHabu) [reply flytekit/common/tasks/presto_task.py:132]: Should we "unify" the setting of those implicit inputs between those two call paths? maybe make __call__ call this function or something?

File: flytekit/common/tasks/task.py (hunk 3 above)
Comment (EngHabu) [reply flytekit/common/tasks/task.py:282]: Is it true that client side task versions do not work with container tasks? why not? I thought that's exactly what we do today... (we grab the Image version and make it the task version)...

**Comment by katrogan** [comment]
> Is there an example execution of single task from command line / jupyter?

@kumare3 adding docs separately :) https://github.com/lyft/flyte/pull/354

File: flytekit/common/tasks/presto_task.py (hunk 1 above)
Comment (katrogan) [reply flytekit/common/tasks/presto_task.py:132]: `__call__` only creates bindings for inputs, rather than converting them to sdk types but good point i can refactor the behavior to a shared function

File: flytekit/common/tasks/task.py (hunk 4 above)
Comment (katrogan) [reply flytekit/common/tasks/task.py:307]: done

File: flytekit/common/tasks/task.py (hunk 3 above)
Comment (katrogan) [reply flytekit/common/tasks/task.py]: thanks for the catch/suggestion! done

**Comment by kumare3** [comment]
@katrogan do we want to add the example notebook that you had to this? Or maybe I will just update my raw container task example to use your single task execution :)

**Comment by katrogan** [comment]
@kumare3 do we have a designated place in the docs or flytekit for adding example notebooks?

**Comment by kumare3** [comment]
> @kumare3 do we have a designated place in the docs or flytekit for adding example notebooks?

I am actually creating one in my Raw Container PR. For now I think you can continue merge it, I will use it in my PR

File: flytekit/models/common.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/models/common.py:59]: yeah git sha is used for task version, this is just used for equality.

File: flytekit/common/mixins/registerable.py [hunk 6]
```diff
@@ -153,10 +153,18 @@ def some_task()
...
+                    self._platform_valid_name = _utils.fqdn(m.__name__, k, entity_type=self.resource_type)
+                    _logging.debug("Auto-assigning name to {}".format(self._platform_valid_name))
+                    return
+            except ValueError:
+                # Empty pandas dataframes behave weirdly here such that calling `m.df` raises:
+                # ValueError: The truth value of a {type(self).__name__} is ambiguous. Use a.empty, a.bool(), a.item(),
+                #   a.any() or a.all()
+                # Since dataframes aren't registrable entities to begin with we swallow any errors they raise and
+                # continue looping through m.
+                pass
```
Comment (wild-endeavor) [inline flytekit/common/mixins/registerable.py:167]: Can you add a debugging log here just in case.

File: flytekit/common/tasks/task.py (hunk 4 above)
Comment (wild-endeavor) [reply flytekit/common/tasks/task.py:307]: I think revamping everything to use Python 3 typing will be a separate effort - one that will probably take a while to do.  We can incrementally add along the way sure, either way, no preference.

File: flytekit/common/tasks/task.py [hunk 7]
```diff
@@ -268,6 +273,59 @@ def _python_std_input_map_to_literal_map(self, inputs):
             for k, v in _six.iteritems(self.interface.inputs)
         })
 
+    def _produce_deterministic_version(self, version=None):
+        """
+        :param Text version:
+        :return Text:
+        """
+
+        if self.container is not None and self.container.data_config is None:
```
Comment (wild-endeavor) [inline flytekit/common/tasks/task.py:282]: so no white list?  can you add a comment explaining why data config can't be none?

File: flytekit/common/mixins/registerable.py (hunk 6 above)
Comment (katrogan) [reply flytekit/common/mixins/registerable.py:167]: done

File: flytekit/common/tasks/task.py (hunk 7 above)
Comment (katrogan) [reply flytekit/common/tasks/task.py:282]: done. i'd rather not add a whitelist because barring the raw container task anything that's containerless (e.g. Sagemaker) in the future should just seamlessly be able to call this function without a code change here
